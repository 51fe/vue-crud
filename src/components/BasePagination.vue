<template>
  <div
    :class="{'hidden':hidden}"
    class="pagination-container"
  >
    <el-pagination
      :current-page="page"
      :page-size="limit"
      :layout="layout"
      :page-sizes="pageSizes"
      :total="total"
      :background="background"
      @size-change="handleSizeChange"
      @current-change="handleCurrentChange"
    />
  </div>
</template>

<script>
import { scrollTo } from '../utils/scroll-to'

export default {
  name: 'BasePagination',
  props: {
    total: {
      type: Number,
      default: 0
    },
    page: {
      type: Number,
      default: 1
    },
    limit: {
      type: Number,
      default: 15
    },
    pageSizes: {
      type: Array,
      default: () => [15, 30, 50, 100]
    },
    layout: {
      type: String,
      default: 'total, sizes, prev, pager, next, jumper'
    },
    background: {
      type: Boolean,
      default: false
    },
    autoScroll: {
      type: Boolean,
      default: true
    },
    hidden: {
      type: Boolean,
      default: false
    }
  },
  methods: {
    handleSizeChange(val) {
      this.$emit('update:limit', val)
      this.$emit('pagination', { page: this.page, limit: val })
      if (this.autoScroll) {
        scrollTo(0, 800)
      }
    },
    handleCurrentChange(val) {
      this.$emit('update:page', val)
      this.$emit('pagination', { page: val, limit: this.limit })
      if (this.autoScroll) {
        scrollTo(0, 800)
      }
    }
  }
}
</script>

<style>
.pagination-container {
  display: flex;
  clear: both;
  background: #fff;
}
.pagination-container .el-pagination{
  margin: 8px auto 24px auto;
}
.pagination-container.hidden {
  display: none;
}
</style>
