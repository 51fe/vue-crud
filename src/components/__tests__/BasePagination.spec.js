import { fireEvent, render, screen } from '@testing-library/vue'
import BasePagination from '../BasePagination.vue'

const setUp = () => {
  return render(BasePagination, {
    props: {
      total: 32
    }
  })
}

test('renders correctly by default', () => {
  const { html } = setUp()
  screen.getByText(/共 32 条/i)
  expect(html()).toMatchSnapshot()
})

test('emits events correctly when click the pager number', async() => {
  const page = 2
  const { emitted } = setUp()
  const numberBtn = screen.getByText(page, { selector: '.number' })
  // click page 2 button
  await fireEvent.click(numberBtn)
  expect(emitted()['update:page'][0][0]).toBe(page)
  expect(emitted()['pagination'][0][0].page).toBe(page)
})

test('emits events correctly when click the page size item', async() => {
  const limit = 30
  const { emitted } = setUp()
  const item = screen.getByText(new RegExp(`${limit}条/页`))
  // click the second item
  await fireEvent.click(item)
  expect(emitted()['update:limit'][0][0]).toBe(limit)
  expect(emitted()['pagination'][0][0].limit).toBe(limit)
})


