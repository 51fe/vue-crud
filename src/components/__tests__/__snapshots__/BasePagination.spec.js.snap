// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly by default 1`] = `
<div class="pagination-container">
  <div class="el-pagination"><span class="el-pagination__total">共 32 条</span><span class="el-pagination__sizes"><div class="el-select el-select--mini"><!----><div class="el-input el-input--mini el-input--suffix"><!----><input type="text" readonly="readonly" autocomplete="off" placeholder="请选择" class="el-input__inner"><!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----><!----><!----><!----></span>
    <!----></span>
    <!---->
    <!---->
  </div>
  <transition-stub name="el-zoom-in-top">
    <div class="el-select-dropdown el-popper" style="display: none;">
      <div class="el-scrollbar" style="display: none;">
        <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
          <ul class="el-scrollbar__view el-select-dropdown__list">
            <!---->
            <li class="el-select-dropdown__item selected"><span>15条/页</span></li>
            <li class="el-select-dropdown__item"><span>30条/页</span></li>
            <li class="el-select-dropdown__item"><span>50条/页</span></li>
            <li class="el-select-dropdown__item"><span>100条/页</span></li>
          </ul>
        </div>
        <div class="el-scrollbar__bar is-horizontal">
          <div class="el-scrollbar__thumb" style="width: 0px; transform: translateX(0%); webkit-transform: translateX(0%);"></div>
        </div>
        <div class="el-scrollbar__bar is-vertical">
          <div class="el-scrollbar__thumb" style="height: 0px; transform: translateY(0%); webkit-transform: translateY(0%);"></div>
        </div>
      </div>
      <p class="el-select-dropdown__empty">
        无数据
      </p>
    </div>
  </transition-stub>
</div></span><button type="button" disabled="disabled" class="btn-prev"><i class="el-icon el-icon-arrow-left"></i></button>
<ul class="el-pager">
  <li class="number active">1</li>
  <!---->
  <li class="number">2</li>
  <!---->
  <li class="number">3</li>
</ul><button type="button" class="btn-next"><i class="el-icon el-icon-arrow-right"></i></button><span class="el-pagination__jump">前往<div class="el-input el-pagination__editor is-in-pagination"><!----><input type="number" autocomplete="off" min="1" max="3" class="el-input__inner"><!----><!----><!----><!----></div>页</span></div>
</div>
`;
